hipercam/scripts/shiftadd.py

```diff
@@ -0,0 +1,663 @@
+            yoff += dy
+            offsets.append((xoff, yoff))
+
+        # offsets are now defined w.r.t to the positions in the
+        # aperture file for first image, and then w.r.t to
+        # previous file for subsequent images. For each CCD,
+        # we should center the offsets on 0,0. This minimises the
+        # risks of having unsampled pixels in the full frame data
+        offsets = np.array(offsets)
+        offsets -= offsets.mean(axis=0)
+
+        output_mccd = hcam.MCCD.read(str(files[0].strip()))
+        cnames = sorted(list(output_mccd.keys()))
+        header_string = "nframes="
+
+        for cnam in cnames:
+            print(f"stacking CCD {cnam}")
+
+            try:
+                # make real celestial WCS from header
+                orig_wcs = wcs_from_header(output_mccd)
+            except Exception as err:
+                # fallback to basic WCS which can be used for any method except 'exact'
+                if reprmethod == "exact":
+                    raise hcam.HipercamError(
+                        "failed to create WCS from header, cannot use 'exact' reprojection method"
+                    ) from err
+                # start with basic WCS, CRDELT1, no offsets
+                orig_wcs = wcs.WCS(naxis=2)
+
+            arrs = []
+            nframes_used = 0
+            for ifile, fname in enumerate(files):
+                fname = str(fname.strip())
+                mccd = hcam.MCCD.read(fname)
+                ccd = mccd[cnam]
+                if not ccd.is_data():
+                    continue
+
+                # skip if FWHM is above threshold
+                if fthresh > 0 and fwhm_values[ifile] > fthresh:
+                    print(
+                        f"skipping frame {ifile + 1} (FWHM too large ({fwhm_values[ifile]:.1f} > {fthresh:.1f}))"
+                    )
+                    continue
+
+                print("resampling frame", ifile + 1)
+                nframes_used += 1
+
+                # find offset from previous frame
+                cumulative_offset_x, cumulative_offset_y = offsets[ifile]
+
+                # find output shape to resample each window onto (binned)
+                # full frame array
+                output_shape = (ccd.nytot // ccd.head.ybin, ccd.nxtot // ccd.head.xbin)
+
+                # add this CCD to the list of arrays
+                for wnam, wind in ccd.items():
+                    # make WCS to reproject data onto full frame
+                    # (binned) array
+                    pixel_wcs = copy.deepcopy(orig_wcs)
+                    crval1, crval2 = pixel_wcs.wcs.crpix
+                    window_offset_x = (wind.llx - 1) / wind.xbin
+                    window_offset_y = (wind.lly - 1) / wind.ybin
+                    pixel_wcs.wcs.crpix = (
+                        crval1 + cumulative_offset_x - window_offset_x,
+                        crval2 + cumulative_offset_y - window_offset_y,
+                    )
```

Comment:
Incidentally, having been messing around with trying to plot the non-reprojected window over a reprojected one I'm not certain if the `-1`s here are right. Most obviously in the case where the offsets are (0,0) for a "base" frame you just end up moving the data 1 pixel towards the lower left for seemingly no reason.

Reply:
Good idea to test with 0 offset. There could well be a fencepost error here, it's something I'm particularly prone to.